src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -0,0 +1,276 @@
+  }
+
+  private def persistStatus(call: Call, callStatus: CallStatus): Future[Unit] = persistStatus(Iterable(call), callStatus)
+
+  private def persistStatus(calls: Traversable[Call], callStatus: CallStatus): Future[Unit] = {
+    executionStore ++= calls.map { _ -> callStatus}.toMap
+    log.info("Before status persist " + callStatus + " for calls " + calls.map(_.fullyQualifiedName).mkString(", "))
+    val future = dataAccess.setStatus(workflow.id, calls, callStatus)
+    log.info("After async status persist " + callStatus)
+    future
+  }
+
+  private def handleCallCompleted(call: Call, outputs: CallOutputs): Future[Unit] = {
+    for {
+      _ <- dataAccess.setOutputs(workflow.id, call, outputs)
+      _ <- persistStatus(call, Done)
+    } yield ()
+  }
+
+  private def startActor(call: Call, locallyQualifiedInputs: Map[String, WdlValue]): Unit = {
+    val callActorProps = CallActor.props(call, locallyQualifiedInputs, backend, workflow, "CallActor-" + call.name)
+    log.info("Before actor creation for " + call.fullyQualifiedName)
+    val callActor = context.actorOf(callActorProps)
+    log.info("Before actor messaging for " + call.fullyQualifiedName)
+    callActor ! CallActor.Start
+    log.info("Started actor for " + call.fullyQualifiedName)
+  }
+
+  /**
+   * Start all calls which are currently in state `NotStarted` and whose prerequisites are all `Done`,
+   * i.e. the calls which should now be eligible to run.  Update status to `Starting` in DataAccess and
+   * start Call actors.  Return a Future of Iterable[Unit] to allow the caller to compose or block.
+   */
+  type CallAndInputs = (Call, Map[String, WdlValue])
+  private def startRunnableCalls(): Unit = {
+
+    def isRunnable(call: Call) = call.prerequisiteCalls().forall(executionStore.get(_).get == ExecutionStatus.Done)
+    /**
+     * Start all calls which are currently in state `NotStarted` and whose prerequisites are all `Done`,
+     * i.e. the calls which should now be eligible to run.
+     */
+    def findRunnableCalls: Iterable[Call] = {
+      for {
+        callEntry <- executionStore if callEntry._2 == ExecutionStatus.NotStarted
+        call = callEntry._1 if isRunnable(call)
+      } yield call
+    }
+
+    val runnableCalls = findRunnableCalls
+    if (runnableCalls.isEmpty) {
+      log.info("No runnable calls to start.")
+    } else {
+      log.info("Starting calls: " + runnableCalls.map {_.name}.toSeq.sorted.mkString(", "))
+      val futureCallsAndInputs = for {
+        _ <- persistStatus(runnableCalls, Starting)
+        allInputs <- Future.sequence(runnableCalls map fetchLocallyQualifiedInputs)
+      } yield runnableCalls zip allInputs
+      val callsAndInputs = Await.result(futureCallsAndInputs, DatabaseTimeout)
+      log.info("Calls and inputs are " + callsAndInputs)
+      callsAndInputs foreach { case (call, inputs) => startActor(call, inputs)}
+    }
+  }
+
+  // FIXME FIXME FIXME this needs to be refactored but I don't understand it.
```

Comment:
This function is a bag of hurt. I'm surprised that Odersky hasn't taken away our scala license. Perhaps I've found something to do in the car on my trip? :)